internal/volumes.go

```diff
@@ -15,278 +16,157 @@ var (
+type IRODSFSPathMapping struct {
+	IRODSPath    string `yaml:"irods_path" json:"irods_path"`
+	MappingPath  string `yaml:"mapping_path" json:"mapping_path"`
+	ResourceType string `yaml:"resource_type" json:"resource_type"` // file or dir
 }
 
-func (i *Internal) outputCSIVolumeHandle(job *model.Job, mountID int) string {
-	return fmt.Sprintf("%s-handle-%s-%d", csiDriverOutputVolumeNamePrefix, job.InvocationID, mountID)
+func (i *Internal) getCSIVolumeHandle(job *model.Job) string {
+	return fmt.Sprintf("%s-handle-%s", csiDriverVolumeNamePrefix, job.InvocationID)
 }
 
-func (i *Internal) inputCSIVolumeName(job *model.Job, mountID int) string {
-	return fmt.Sprintf("%s-%s-%d", csiDriverInputVolumeNamePrefix, job.InvocationID, mountID)
+func (i *Internal) getCSIVolumeName(job *model.Job) string {
+	return fmt.Sprintf("%s-%s", csiDriverVolumeNamePrefix, job.InvocationID)
 }
 
-func (i *Internal) outputCSIVolumeName(job *model.Job, mountID int) string {
-	return fmt.Sprintf("%s-%s-%d", csiDriverOutputVolumeNamePrefix, job.InvocationID, mountID)
+func (i *Internal) getCSIVolumeClaimName(job *model.Job) string {
+	return fmt.Sprintf("%s-%s", csiDriverVolumeClaimNamePrefix, job.InvocationID)
 }
 
-func (i *Internal) inputCSIVolumeClaimName(job *model.Job, mountID int) string {
-	return fmt.Sprintf("%s-%s-%d", csiDriverInputVolumeClaimNamePrefix, job.InvocationID, mountID)
-}
-
-func (i *Internal) outputCSIVolumeClaimName(job *model.Job, mountID int) string {
-	return fmt.Sprintf("%s-%s-%d", csiDriverOutputVolumeClaimNamePrefix, job.InvocationID, mountID)
-}
+func (i *Internal) getInputPathMappings(job *model.Job) ([]IRODSFSPathMapping, error) {
+	mappings := []IRODSFSPathMapping{}
+	// mark if the mapping path is already occupied
+	// key = mount path, val = irods path
+	mappingMap := map[string]string{}
 
-func (i *Internal) getInputIRODSPaths(job *model.Job) []string {
-	paths := []string{}
 	for _, step := range job.Steps {
 		for _, stepInput := range step.Input {
 			irodsPath := stepInput.IRODSPath()
 			if len(irodsPath) > 0 {
-				paths = append(paths, irodsPath)
-			}
-		}
-	}
-	return paths
-}
 
-func (i *Internal) getOutputIRODSPaths(job *model.Job) []string {
-	paths := []string{}
-	paths = append(paths, job.OutputDirectory())
-
-	/*
-		for _, step := range job.Steps {
-			for _, stepOutput := range step.Output {
-				irodsPath := stepOutput.Source()
-				if len(irodsPath) > 0 {
-					paths = append(paths, irodsPath)
+				resourceType := "file"
+				if stepInput.Type == "FileInput" || strings.ToLower(stepInput.Type) == "fileinput" {
+					resourceType = "file"
+				} else if stepInput.Type == "MultiFileSelector" || strings.ToLower(stepInput.Type) == "multifileselector" {
+					resourceType = "file"
+				} else if stepInput.Type == "FolderInput" || strings.ToLower(stepInput.Type) == "folderinput" {
+					resourceType = "dir"
+				} else {
+					// unknown
+					return nil, fmt.Errorf("unknown step input type - %s", stepInput.Type)
```

Comment:
For cases where the first condition is matched by a subset of values matched by the second condition, I think I'd omit the first condition:

```suggestion
				resourceType := "file"
				if strings.ToLower(stepInput.Type) == "fileinput" {
					resourceType = "file"
				} else if strings.ToLower(stepInput.Type) == "multifileselector" {
					resourceType = "file"
				} else if strings.ToLower(stepInput.Type) == "folderinput" {
					resourceType = "dir"
				} else {
					// unknown
					return nil, fmt.Errorf("unknown step input type - %s", stepInput.Type)
```

Reply:
Applied.